### ISSUE TITLE
Postgres service points to bad data directory after #109075

### ISSUE THREAD

**Issue by indirect** [Homebrew/homebrew-core#109644]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
❯ brew config
HOMEBREW_VERSION: 3.5.10-118-g9972cba                                                   
ORIGIN: https://github.com/Homebrew/brew                                                
HEAD: 9972cbaba233bdfd4c1b60cc2c2dea0cd5d3e03b                 
Last commit: 22 hours ago                                                               
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core                            
Core tap HEAD: b0dd4e755f781cdb0160c716a3b49bf6d9ae136f
Core tap last commit: 25 minutes ago
Core tap branch: master    
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_BROWSER: open                                                                  
HOMEBREW_CASK_OPTS: []                     
HOMEBREW_EDITOR: mvim -f --nomru -c "au VimLeave * !open -b $__CFBundleIdentifier"
HOMEBREW_MAKE_JOBS: 10                     
Homebrew Ruby: 2.6.8 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 13.1.6 build 1316                 
Git: 2.37.1 => /opt/homebrew/bin/git       
Curl: 7.84.0 => /usr/bin/curl
macOS: 13.0-arm64                          
CLT: 14.0.0.0.1.1659797938
Xcode: 13.4.1                         
Rosetta 2: false
```

```
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You are using macOS 13.
We do not provide support for this pre-release version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
pre-release version.
```


### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Use postgresql, installed via `brew install postgresql@14` followed by `brew services restart postgresql@14`.

### What happened (include all command output)?

I ran `brew install postgresql`:

```
❯ brew install postgresql
[many other formula were upgraded first]
==> Installing postgresql@14            
==> Pouring postgresql@14--14.5_2.arm64_monterey.bottle.tar.gz
==> Caveats                                                                             
Previous versions of postgresql shared the same data directory.    

You can migrate to a versioned data directory by running:
  mv -v "/opt/homebrew/var/postgres" "/opt/homebrew/var/postgresql@14"
                                                                                        
(Make sure PostgreSQL is stopped before executing this command)

This formula has created a default database cluster with:
  initdb --locale=C -E UTF-8 /opt/homebrew/var/postgres
For more details, read:              
  https://www.postgresql.org/docs/14/app-initdb.html                                                                                                                             
                                                                                                                                                                                 
To restart postgresql@14 after an upgrade:                                                                                                                                       
  brew services restart postgresql@14                                                   
Or, if you don't want/need a background service you can just run:
  /opt/homebrew/opt/postgresql@14/bin/postgres -D /opt/homebrew/var/postgres
```

I followed the instructions to migrate the versioned data directory:

```
[verified postgres is not running first via ps]
❯ mv -v "/opt/homebrew/var/postgres" "/opt/homebrew/var/postgresql@14"                                                                                                           
/opt/homebrew/var/postgres -> /opt/homebrew/var/postgresql@14                           
❯ brew services restart postgresql@14                                                   
==> Successfully started `postgresql@14` (label: homebrew.mxcl.postgresql@14)           
```

The service did not, in fact, start successfully:

```
❯ brew services                                                                                                                                                                  
Name            Status     User  File                                                                                                                                            
postgresql@14   error  512 andre ~/Library/LaunchAgents/homebrew.mxcl.postgresql@14.plist
```

The reason the service is not starting successfuly (I think) is that the plist has the wrong data directory:

```
❯ cat ~/Library/LaunchAgents/homebrew.mxcl.postgresql@14.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
        <key>KeepAlive</key>
        <true/>
        <key>Label</key>
        <string>homebrew.mxcl.postgresql@14</string>
        <key>LimitLoadToSessionType</key>
        <array>
                <string>Aqua</string>
                <string>Background</string>
                <string>LoginWindow</string>
                <string>StandardIO</string>
                <string>System</string>
        </array>
        <key>ProgramArguments</key>
        <array>
                <string>/opt/homebrew/opt/postgresql@14/bin/postgres</string>
                <string>-D</string>
                <string>/opt/homebrew/var/postgres</string>
        </array>
        <key>RunAtLoad</key>
        <true/>
        <key>StandardErrorPath</key>
        <string>/opt/homebrew/var/log/postgresql@14.log</string>
        <key>StandardOutPath</key>
        <string>/opt/homebrew/var/log/postgresql@14.log</string>
        <key>WorkingDirectory</key>
        <string>/opt/homebrew</string>
</dict>
</plist>
```

### What did you expect to happen?

I expected `brew services restart postgres@14` to start postgres. I also expected the output "started successfully" to mean that the daemon was running, not crashed.

Manually editing the plist at `~/Library/LaunchAgents/homebrew.mxcl.postgresql@14.plist` to change the line `<string>/opt/homebrew/var/postgres</string>` to instead read `<string>/opt/homebrew/var/postgresql@14</string>` fixed the service.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
# start with an existing postgres installation
brew install postgres@14
mv -v "/opt/homebrew/var/postgres" "/opt/homebrew/var/postgresql@14"
brew services restart postgres@14
brew services
# the postgres service is now broken, pointing to a directory that doesn't exist
```

**Comment by indirect** [comment]
I would also recommend that you add to the caveats that every Ruby app using the `pg` gem now needs to run `bundle pristine pg` to recompile their postgres bindings against the new directory for libpg.

**Comment by OJFord** [comment]
> I would also recommend that you add to the caveats that every Ruby app using the pg gem now needs to run bundle pristine pg to recompile their postgres bindings against the new directory for libpq.

Not ruby-specific of course, anything using libpq.

**Comment by NewAlexandria** [comment]
Commenting that this worked for me - want to raise more visibility for it.

**Comment by mcls** [comment]
I experienced the same issue. As @indirect said, editing the `.plist` manually works. Alternatively, reinstalling the formula after renaming the directory also worked for me. 

I'm not very familiar with Homebrew internals, but the issue seems to be that when the `old_postgres_data_dir` still exists at install time, it is used that to create the `.plist` file. 

Related code:

https://github.com/Homebrew/homebrew-core/blob/8060df9169055ee1fb9d1d69fd149c3922e864ba/Formula/postgresql%4014.rb#L99-L105

https://github.com/Homebrew/homebrew-core/blob/8060df9169055ee1fb9d1d69fd149c3922e864ba/Formula/postgresql%4014.rb#L155-L161

**Comment by brennacodes** [comment]
+1. Running `bundle pristine pg` fixed this as recommended by @indirect above.

**Comment by Hermanverschooten** [comment]
Another issue I already noticed on 2 different (M1) Macs, is that after the upgrade, `brew services` lists the `postgresql@14` service as failed, while postgres does run. It only is not manageable anymore. 
On my mini this was solved with a reboot, on my air not even that helped, I had to remove the `postgres.plist` from `LaunchAgents`.

Following the directions to move the data directory necessitated editing the plist to point it to that folder.

**Comment by SMillerDev** [comment]
A reinstall and service restart will regenerate the plist too.

**Comment by davidalejandroaguilar** [comment]
Heads up, when following @indirect's fix, you have to edit:

```
/opt/homebrew/Cellar/postgresql@14/14.5_4/homebrew.mxcl.postgresql@14.plist
```

Not: 

```
~/Library/LaunchAgents/homebrew.mxcl.postgresql@14.plist
```

Otherwise it will be re-generated when you `brew restart`.

**Comment by ionescu77** [comment]
Hi, having the same issue. Thank you all for the info.

It's infurating :)

```
brew services
Name          Status     User      File
postgresql@12 none
postgresql@13 none
postgresql@14 error  512 username ~/Library/LaunchAgents/homebrew.mxcl.postgresql@14.plist
```

I've edited the `LaunchAgents/homebrew.mxcl.postgresql@14.plist` but it gets overwriten.
Have found @davidalejandroaguilar comment above ⤴️ (my Cellar has another location):
```
/usr/local/opt/postgresql@14/homebrew.mxcl.postgresql@14.plist
```

And of course also updated here (from `/usr/local/var/postgres` to `/usr/local/var/postgresql@14`)

```xml
	<array>
		<string>/usr/local/opt/postgresql@14/bin/postgres</string>
		<string>-D</string>
		<string>/usr/local/var/postgresql@14</string>
	</array>
```

But still in the log:

```
tail -f /usr/local/var/log/postgresql@14.log

postgres: could not access directory "/usr/local/var/postgres": No such file or directory
Run initdb or pg_basebackup to initialize a PostgreSQL data directory.
postgres: could not access directory "/usr/local/var/postgres": No such file or directory
Run initdb or pg_basebackup to initialize a PostgreSQL data directory.
```

However in this directory `/usr/local/var/postgresql@14` I've found a `postmaster.opts` file having the wrong option still:
```
cat /usr/local/var/postgresql@14/postmaster.opts
/usr/local/Cellar/postgresql/14.0/bin/postgres "-D" "/usr/local/var/postgres"
```

I changed it here too and it still does not work.

I am not sure where the old path is still coming from :(

**Comment by 907th** [comment]
Hello, guys!

I updated the PostgreSQL to the version 14 and moved the data directory with `mv -v "/opt/homebrew/var/postgres" "/opt/homebrew/var/postgresql@14"` (like installation script proposed me to do). And I also found that the service can not start now because its configuration points to the wrong directory.

This is how I resolved the issue:
1. Uninstall PostgreSQL with `brew uninstall postgresql@14`.
2. Unload the service definition with `brew services unload postgresql@14`.
3. Run `brew services info postgresql@14` to make sure that the service is not loaded, is not running and is not schedulable (if it is you may need to run some another commands).
4. Install PostgreSQL with `brew install postgresql@14`.
5. Load the service with `brew services load postgresql@14`.
6. Check that service configuration points to the proper directory now `less ~/Library/LaunchAgents/homebrew.mxcl.postgresql@14.plist`.
7. View the log file `less /usr/local/var/log/postgresql@14.log` and find that the `postmaster.pid` exists and prevents the server from starting.
8. Delete `postmaster.pid` manually with `rm /usr/local/var/postgresql@14/postmaster.pid`.
9. Wait a little and check that PostgreSQL is running now!

I hope this will help someone!

**Comment by sainttttt** [comment]
> 8\. Delete `postmaster.pid` manually with `rm /usr/local/var/postgresql@14/postmaster.pid`.

Thanks, your instructions worked for me! Only difference was that I had to do `rm /usr/local/var/postgresql/postmaster.pid` instead of `rm /usr/local/var/postgresql@14/postmaster.pid`

**Comment by adelosa** [comment]
Encountered the same issue - followed @907th instructions to get running. Observed same issue where the plist file was pointing to the old non versioned data directory.

**Comment by Hermanverschooten** [comment]
Best solution is to uninstall/reinstall with the moving of the data folder in between. I ran into issues when trying to upgrade to 15 on the Mac where I did the manual changes, but not on the one where I reinstalled.
